core/services/relay/evm/chain_reader.go

```diff
@@ -169,53 +183,55 @@ func addEncoderDef(name string, method abi.Method, parsed *parsedTypes, chainRea
 		return err
 	}
 	input.mod = inputMod
-	parsed.encoderDefs[wrapItemType(name, true)] = input
+	parsed.encoderDefs[wrapItemType(contractName, typeName, true)] = input
 	return nil
 }
 
-func addDecoderDef(name string, outputs abi.Arguments, parsed *parsedTypes, def types.ChainReaderDefinition) error {
+func addDecoderDef(contractName, methodName string, outputs abi.Arguments, parsed *parsedTypes, def types.ChainReaderDefinition) error {
 	output := &codecEntry{Args: outputs}
 	mod, err := def.OutputModifications.ToModifier(evmDecoderHooks...)
 	if err != nil {
 		return err
 	}
 	output.mod = mod
-	parsed.decoderDefs[wrapItemType(name, false)] = output
+	parsed.decoderDefs[wrapItemType(contractName, methodName, false)] = output
 	return output.Init()
 }
 
-func addTypes(chainContractReaders map[string]types.ChainContractReader, parsed *parsedTypes) (map[string]common.Hash, error) {
-	events := map[string]common.Hash{}
+func addTypes(chainContractReaders map[string]types.ChainContractReader, parsed *parsedTypes) (map[string]map[string]common.Hash, error) {
+	events := map[string]map[string]common.Hash{}
 	for contractName, chainContractReader := range chainContractReaders {
+		eventMap := map[string]common.Hash{}
+		events[contractName] = eventMap
 		contractAbi, err := abi.JSON(strings.NewReader(chainContractReader.ContractABI))
 		if err != nil {
 			return nil, err
 		}
 
-		for chainReadingDefinitionName, chainReaderDefinition := range chainContractReader.ChainReaderDefinitions {
+		for typeName, chainReaderDefinition := range chainContractReader.ChainReaderDefinitions {
 			switch chainReaderDefinition.ReadType {
 			case types.Method:
-				err = addMethods(chainReadingDefinitionName, contractAbi, chainReaderDefinition, parsed)
+				err = addMethods(contractName, typeName, contractAbi, chainReaderDefinition, parsed)
 			case types.Event:
 				var hash common.Hash
-				hash, err = addEventTypes(chainReadingDefinitionName, contractAbi, chainReaderDefinition, parsed)
-				events[chainReadingDefinitionName] = hash
+				hash, err = addEventTypes(contractName, typeName, contractAbi, chainReaderDefinition, parsed)
+				eventMap[typeName] = hash
```

Comment:
eventMap[typeName], err = addEventTypes(contractName, typeName, contractAbi, chainReaderDefinition, parsed)
```suggestion
				hash, err = addEventTypes(contractName, typeName, contractAbi, chainReaderDefinition, parsed)
				eventMap[typeName] = hash
```